nuxeo-wopi/src/main/java/org/nuxeo/wopi/WOPIServiceImpl.java

```diff
@@ -213,4 +242,81 @@ public boolean verifyProofKey(String proofKeyHeader, String oldProofKeyHeader, S
         }
         return res;
     }
+
+    @Override
+    public boolean refreshDiscovery() {
+        byte[] discoveryBytes = fetchDiscovery();
+        if (ArrayUtils.isEmpty(discoveryBytes)) {
+            return false;
+        }
+
+        log.debug("Successfully fetched WOPI dicovery");
+        storeDiscovery(discoveryBytes);
+        loadDiscovery(discoveryBytes);
+        if (Framework.isBooleanPropertyTrue(CLUSTERING_ENABLED_PROP)) {
+            invalidator.sendMessage(new WOPIDiscoveryInvalidation(discoveryBytes));
+        }
+        return true;
+    }
+
+    protected byte[] fetchDiscovery() {
+        if (discoveryURL == null) {
+            log.warn("No WOPI discovery URL configured, cannot fetch discovery. Please configure the '{}' property.",
+                    WOPI_DISCOVERY_URL_PROPERTY);
+            return ArrayUtils.EMPTY_BYTE_ARRAY;
+        }
+
+        log.debug("Fetching WOPI dicovery from discovery URL {}", discoveryURL);
+        HttpClientBuilder httpClientBuilder = HttpClientBuilder.create();
+        HttpGet request = new HttpGet(discoveryURL);
+        try (CloseableHttpClient httpClient = httpClientBuilder.build();
+                CloseableHttpResponse response = httpClient.execute(request);
+                InputStream is = response.getEntity().getContent()) {
+            return IOUtils.toByteArray(is);
+        } catch (IOException e) {
+            log.error("Error while fetching WOPI discovery: {}", e::getMessage);
+            log.debug(e, e);
+            return ArrayUtils.EMPTY_BYTE_ARRAY;
+        }
+    }
+
+    protected void storeDiscovery(byte[] discoveryBytes) {
+        keyValueStore.put(WOPI_DISCOVERY_KEY, discoveryBytes);
+    }
+
+    public static class WOPIDiscoveryInvalidation implements SerializableMessage {
+
+        private static final long serialVersionUID = 1L;
+
+        public final byte[] discovery;
+
+        public WOPIDiscoveryInvalidation(byte[] discovery) {
```

Comment:
But you don't need to pass the discovery bytes in the invalidation message, it's already available in the key/value store.